jd4/daemon.py

```diff
@@ -24,30 +27,33 @@ def __init__(self, session, request, ws):
         self.request = request
         self.ws = ws
         self.tag = self.request.pop('tag', None)
+        self.event = self.request.pop('event', None)
```

Comment:
Move this event handling out of judge handler. i.e. you can add an EventHandler, though the naming is weird.